chips/stm32f303xc/src/flash.rs

```diff
@@ -0,0 +1,608 @@
+            location.set(halfword);
+
+            self.buffer.replace(buffer);
+        });
+    }
+
+    pub fn erase_page(&self, page_number: usize) -> ReturnCode {
+        if page_number > 128 {
+            return ReturnCode::EINVAL;
+        }
+
+        while self.registers.sr.is_set(Status::BSY) {}
+
+        if self.is_locked() {
+            self.unlock();
+        }
+
+        self.enable();
+        self.state.set(FlashState::Erase);
+
+        // Choose page erase mode.
+        self.registers.cr.modify(Control::PER::SET);
+        self.registers
+            .ar
+            .write(Address::FAR.val((PAGE_START + page_number * PAGE_SIZE) as u32));
+        self.registers.cr.modify(Control::STRT::SET);
+
+        ReturnCode::SUCCESS
+    }
+
+    pub fn erase_all(&self) -> ReturnCode {
+        while self.registers.sr.is_set(Status::BSY) {}
+
+        if self.is_locked() {
+            self.unlock();
+        }
+
+        self.enable();
+        self.state.set(FlashState::Erase);
+
+        // Choose mass erase mode.
+        self.registers.cr.modify(Control::MER::SET);
+        self.registers.cr.modify(Control::STRT::SET);
+
+        ReturnCode::SUCCESS
+    }
+
+    pub fn write_page(
+        &self,
+        page_number: usize,
+        buffer: &'static mut StmF303Page,
+    ) -> Result<(), (ReturnCode, &'static mut StmF303Page)> {
+        if page_number > 128 {
+            return Err((ReturnCode::EINVAL, buffer));
+        }
+
+        while self.registers.sr.is_set(Status::BSY) {}
+
+        if self.is_locked() {
+            self.unlock();
+        }
+
+        self.enable();
+        self.state.set(FlashState::Write);
+
+        // Choose programming mode.
+        self.registers.cr.modify(Control::PG::SET);
+
+        self.buffer.replace(buffer);
+        self.page_number.set(page_number);
+        self.program_halfword();
+
+        Ok(())
+    }
+
+    pub fn read_page(
+        &self,
+        page_number: usize,
+        buffer: &'static mut StmF303Page,
+    ) -> Result<(), (ReturnCode, &'static mut StmF303Page)> {
+        if page_number > 128 {
+            return Err((ReturnCode::EINVAL, buffer));
+        }
+
+        while self.registers.sr.is_set(Status::BSY) {}
+
+        let mut byte: *const u8 = (PAGE_START + page_number * PAGE_SIZE) as *const u8;
+        unsafe {
+            for i in 0..buffer.len() {
+                buffer[i] = *byte;
+                byte = byte.offset(1);
+            }
+        }
+
+        self.buffer.replace(buffer);
+        self.state.set(FlashState::Read);
+        DEFERRED_CALL.set();
+
+        Ok(())
+    }
+
+    /// Allows programming the 8 option bytes:
+    /// 0: RDP, 1: USER, 2: DATA0, 3:DATA1, 4. WRP0, 5: WRP1, 6.WRP2, 7. WRP3
+    pub fn write_option(&self, byte_number: usize, value: u8) -> ReturnCode {
+        if byte_number > 7 {
+            return ReturnCode::EINVAL;
+        }
+
+        while self.registers.sr.is_set(Status::BSY) {}
```

Comment:
Tock tries to avoid blocking interfaces when possible (see https://github.com/tock/tock/blob/master/doc/Design.md#split-phase-operation), with an exception for cases where a datasheet indicates the cycles spent blocking will be less than the cycles it would take to set up and process a callback. Can you provide a pointer here on how long this blocking is expected to last, or why a split-phase design is infeasible? Same for other instances of spinning on the status bit.

Reply:
At first, I implemented all the operations with deferred calls. In that case it was necessary to wait for the BSY bit. It was just that after switching to hardware interrupts, I forgot get to delete them. I haven't found any mention in the datasheet about time spent blocking being less than the cycles need to process the callback. After testing the code without the while loops, I am getting the same results, so I suppose I will just delete them. Thanks for the heads-up.